### PR TITLE
Feature/hw 1 hangman

### PR DESCRIPTION
**Pull request by dmytrochornyy**
Asking to check my history one more time

### CODE REVIEW

File: hangman/hangman.py [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+             that has already been revealed.
+
+    '''
+    #creating boolean list with length of wordlist( if match = True), than apply this list to wordlist and print it
+    print(*[i for (i, v) in zip(wordlist, list(map(lambda x: match_with_gaps(my_word, x), wordlist))) if v])
+    
+
+
+
+def hangman_with_hints(secret_word):
```
Comment (Hozhyi) [inline hangman/hangman.py:249]: The code of this function and function hangman are duplicating each other in many terms. Try to make 1 function which can be used in both ways

File: hangman/hangman.py [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+
+if __name__ == "__main__":
+    # pass
+
+    # To test part 2, comment out the pass line above and
+    # uncomment the following two lines.
+    
+    #secret_word = choose_word(wordlist)
+    #hangman(secret_word)
+    #show_possible_matches("t_ _ t")
```
Comment (Hozhyi) [inline hangman/hangman.py:354]: Commented code - is a dead one, there's no sense to have it, just remove

**Review by Hozhyi — CHANGES_REQUESTED** [review]
That was only overall look

**Comment by dmytrochornyy** [comment]
Hi! I`ve fully rebuild my game! Could you look at it one more time, please? @Hozhyi

File: hangman/hangman.py [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+    '''
+    my_word: string with _ characters, current guess of secret word
+    other_word: string, regular English word
+    returns: boolean, True if all the actual letters of my_word match the 
+        corresponding letters of other_word, or the letter is the special symbol
+        _ , and my_word and other_word are of the same length;
+        False otherwise: 
+    '''
+    #delete all spaces, to work more convenient
+    my_word = my_word.replace(" ", "")
```
Comment (Hozhyi) [inline hangman/hangman.py:235]: You could delete them on 1 level up, as you call this function many time with the same `my_word`

File: hangman/hangman.py [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+
+    #check if len is same
+    if len(my_word)!=len(other_word):
+      return False
+    
+    #if len is same
+    else:
```
Comment (Hozhyi) [inline hangman/hangman.py:242]: I think you don't need this `else` statement here, as `return` in scope of previous `if` will return from this function

File: hangman/hangman.py [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+      for i in range(len(my_word)):
+
+        #check if letters on same positions are different (except positions with"_")
+        if my_word[i] != "_" and my_word[i] != other_word[i]:
```
Comment (Hozhyi) [inline hangman/hangman.py:246]: Have you taken into account that result for the following two words should be `False`:
a_bc
abbc

File: hangman/hangman.py [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+    while guesses_remaining > 0 and not is_word_guessed(secret_word, letters_guessed):
+
+        #printing message about guesses remaining
+        print_guesses_remaining(letters_guessed, guesses_remaining)
+
+        #asking for new letter
+        letter = input("Please guess a letter: ").lower()
+        
+        #checking if this is really letter, not another symbol, then update remaining warnings and guesses
+        guesses_remaining, warnings_remaining, letter_is_correct, msg = letter_checker(letter, letters_guessed, guesses_remaining, warnings_remaining)
```
Comment (Hozhyi) [inline hangman/hangman.py:302]: This is quite long line, try to split it into 2 lines of code

File: hangman/hangman.py [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+        
+        #activating hints
+        if hints_on and letter == "*":
```
Comment (Hozhyi) [inline hangman/hangman.py:305]: It seems like you reduce number of warnings when user ask for hints, isn't it?

File: hangman/hangman.py [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+  guesses_remaining - integer of remaining guesses
+  Function prints result of hangman game
+  """
+  # if word was guessed
+  if is_word_guessed(secret_word, letters_guessed):
+    print(
+      "------------\n",
+      "Congratulations, you won! " 
+      "Your total score for this game is: {0}".format(get_score(secret_word, guesses_remaining)),
+      sep=""
```
Comment (Hozhyi) [inline hangman/hangman.py:213]: What is the default value of `sep`?

File: hangman/hangman.py [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+      lowercase
+    letters_guessed: list (of letters), which letters have been guessed so far;
+      assumes that all letters are lowercase
+    returns: boolean, True if all the letters of secret_word are in letters_guessed;
+      False otherwise
+    '''
+    #looking for each letter from secret_word
+    for letter in secret_word:
+      #if at least one letter is not guessed
+      if not letters_guessed[letter]:
```
[the remark this answers is not in the document]
Comment (Hozhyi) [reply hangman/hangman.py:56]: Ah.. I see - you use dictionary (although in the function docstring mentioned that `letters_guessed` is a list object)

File: hangman/hangman.py [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+    hints_on: boolean. True, if you want to play with hints. Default = False
+    guesses_remaining: integer, the number of guesses. Default = 6
+    warnings_remaining: integer, the number of warnings. Default = 3
+    Starts up an interactive game of Hangman.
+    '''
+    #printing hello message
+    print_game_start(secret_word)
+    
+    #create blank letters 0/1-table. 0(False) - if letter is not guesed - 1(True) otherwise
+    letters_guessed = {i: 0 for i in string.ascii_lowercase}
```
Comment (Hozhyi) [inline hangman/hangman.py:290]: It would be more logical for me to use a `set` here, which will represent set of entered letters

**Review by Hozhyi — CHANGES_REQUESTED** [review]

File: hangman/hangman.py [hunk 11]
```diff
@@ -0,0 +1,328 @@
+# Problem Set 2, hangman.py
+# Name: Сhornyy Dmytro
+# Collaborators: Made by myself (Can help others with some ideas)
```
[the remark this answers is not in the document]
Comment (dmytrochornyy) [reply hangman/hangman.py:3]: Done

File: hangman/hangman.py [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+    secret_word: string, the word the user is guessing; assumes all letters are
+      lowercase
+    letters_guessed: list (of letters), which letters have been guessed so far;
+      assumes that all letters are lowercase
+    returns: boolean, True if all the letters of secret_word are in letters_guessed;
+      False otherwise
+    '''
+    #looking for each letter from secret_word
+    for letter in secret_word:
+      #if at least one letter is not guessed
```
[the remark this answers is not in the document]
Comment (dmytrochornyy) [reply hangman/hangman.py:55]: Solved

File: hangman/hangman.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (dmytrochornyy) [reply hangman/hangman.py:56]: Solved

File: hangman/hangman.py [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+    '''
+    #making blank string
+    guessed_word = ''
+
+    for letter in secret_word:
+      #if we already know the letter - add it. If not - add "_ "
+      if letters_guessed[letter]:
+        guessed_word += letter
+      else:
+        guessed_word +="_ "
```
[the remark this answers is not in the document]
Comment (dmytrochornyy) [reply hangman/hangman.py:77]: Solved

File: hangman/hangman.py [hunk 14]
```diff
@@ -0,0 +1,328 @@
...
+    '''
+    letters_guessed: list (of letters), which letters have been guessed so far
+    returns: string (of letters), comprised of letters that represents which letters have not
+      yet been guessed.
+    '''
+    #make emply string
+    available_letters = ''
+
+    #add all unknown letters in cycle
+    for letter in letters_guessed.keys():
```
[the remark this answers is not in the document]
Comment (dmytrochornyy) [reply hangman/hangman.py:91]: Solved

File: hangman/hangman.py [hunk 15]
```diff
@@ -0,0 +1,328 @@
...
+      if not letters_guessed[letter]:
+        available_letters += letter
```
[the remark this answers is not in the document]
Comment (dmytrochornyy) [reply hangman/hangman.py:93]: Solved

File: hangman/hangman.py [hunk 16]
```diff
@@ -0,0 +1,328 @@
...
+def print_game_start(secret_word):
+    """
+    Gets:
+    secret_word:string
+    Prints hello-message
+    """
+    print("Welcome to the game Hangman!",
+    "I am thinking of a word that is {0} letters long.".format(len(secret_word)),
+    sep="\n"
+    )
```
[the remark this answers is not in the document]
Comment (dmytrochornyy) [reply hangman/hangman.py:107]: Solved

File: hangman/hangman.py [hunk 17]
```diff
@@ -0,0 +1,328 @@
...
+
+  #adding new letter to our list of already guessed letters
+  letters_guessed[letter] = 1
+
+  #if guess is good - return good msg
+  if letter in secret_word:
+    return (guesses_remaining, letters_guessed, msg_good)
+
+  #otherwise - return bad msg and substract guesses 
+  elif letter in "aeiou":
```
[the remark this answers is not in the document]
Comment (dmytrochornyy) [reply hangman/hangman.py:181]: Changed to tuple

File: hangman/hangman.py [hunk 18]
```diff
@@ -0,0 +1,328 @@
...
+
+def get_score(secret_word, guesses_remaining):
+  """
+  Gets: 
+  secret_word: string
+  guesses_remaining: integer of remaining guesses
+
+  Function calculates a score of game with given word and remaining guesses and returns it
+  """
+  return guesses_remaining*len(set(secret_word))
```
[the remark this answers is not in the document]
Comment (dmytrochornyy) [reply hangman/hangman.py:196]: Solved

File: hangman/hangman.py (hunk 4 above)
Comment (dmytrochornyy) [reply hangman/hangman.py:242]: You`re right

File: hangman/hangman.py (hunk 5 above)
Comment (dmytrochornyy) [reply hangman/hangman.py:246]: Thank you! Solved.

File: hangman/hangman.py (hunk 6 above)
Comment (dmytrochornyy) [reply hangman/hangman.py:302]: Done with temp_tuple

File: hangman/hangman.py (hunk 7 above)
Comment (dmytrochornyy) [reply hangman/hangman.py:305]: True. Solved.

File: hangman/hangman.py (hunk 8 above)
Comment (dmytrochornyy) [reply hangman/hangman.py:213]: Done

File: hangman/hangman.py [hunk 19]
```diff
@@ -0,0 +1,328 @@
...
+
+
+def print_game_start(secret_word):
+    """
+    Gets:
+    secret_word:string
+    Prints hello-message
+    """
+    print("Welcome to the game Hangman!",
+    "I am thinking of a word that is {0} letters long.".format(len(secret_word)),
```
[the remark this answers is not in the document]
Comment (dmytrochornyy) [reply hangman/hangman.py:105]: Thank you! Forgot about them.

File: hangman/hangman.py (hunk 10 above)
Comment (dmytrochornyy) [reply hangman/hangman.py:290]: as for me, it easier to work with dictionary, for example, it`s enough to write letters_guessed[letter], to check if letter was already guessed. Moreover, as far a know, it faster, to find letter value in dict (letters_guessed[letter]), than find it in tuple (letter in letters_guessed). Im attaching my proof, can you please tell me if i checked speed correctly?

Could not attach file here, I will send it to you in Slack!

File: hangman/hangman.py (hunk 3 above)
Comment (dmytrochornyy) [reply hangman/hangman.py:235]: True. Thank you!

**Comment by dmytrochornyy** [comment]
Hi! I`ve done all changes due to your comments. Could you check this one more time please? @Hozhyi

**Comment by dmytrochornyy** [comment]
Hi! I did all changes that you said. Could you please check that one more time? @Hozhyi

File: hangman/hangman.py [hunk 20]
```diff
@@ -0,0 +1,323 @@
...
+    returns: string (of letters), comprised of letters that represents which letters have not
+      yet been guessed.
+    '''
+    #make emply string
+    available_letters = string.ascii_lowercase
+
+    #add all unknown letters in cycle
+    for letter in set(string.ascii_lowercase):
+        if letter in letters_guessed:
+            available_letters = available_letters.replace(letter, '')
```
Comment (Hozhyi) [inline hangman/hangman.py:86]: Replace function need to look at each letter to check whether this letter needs to be removed. Effectively - it's will be O(n^2). You can make it faster.

File: hangman/hangman.py [hunk 21]
```diff
@@ -0,0 +1,323 @@
...
+        sep="\n"
+        )
+
+
+def letter_checker(letter, letters_guessed, guesses_remaining, warnings_remaining, hints_on):
+    """
+    letter: string, letter to check
+    letters_guessed: set with guessed letters
+    guesses_remaining: int, remaining guesses
+    warnings_remaining: int, remaining warnings
```
Comment (Hozhyi) [inline hangman/hangman.py:124]: Can't see here a description about `hints_on` parameter

File: hangman/hangman.py [hunk 22]
```diff
@@ -0,0 +1,323 @@
...
+    Checks if letter is in appropriate format for guessing
+    returns: tuple with results:(new value of guesses, new value of warnings,
+    boolean - True if word is correct, msg - message to show)
+    """
+    #defining different messages for different cases
+    msg_not_valid = "Oops! That is not a valid letter. You have {} warnings left: {}"
+    msg_repeated = "Oops! You've already guessed that letter. You now have {} warnings: {}"
+    
+    if letter=="*" and hints_on:
```
Comment (Hozhyi) [inline hangman/hangman.py:133]: Could you wrap sign "==" with spaces around?

File: hangman/hangman.py [hunk 23]
```diff
@@ -0,0 +1,323 @@
...
+    secret_word: string
+    letter: string, letter that we are currently guessing
+    letters_guessed: set with guessed letters
+    guesses_remaining: integer of remaining guesses
+    Tryes to guess a letter, returns a tuple: (new int of guesses, new dict of guessed letters, msg to print)
+    """
+    #defining messages to show later
+    msg_good = "Good guess: {}"
+    msg_bad = "Oops! That letter is not in my word: {}"
+    vovels = (i for i in "aeiou")
```
Comment (Hozhyi) [inline hangman/hangman.py:166]: I'd create a separate constant for that variable, and tried to find a better container type

File: hangman/hangman.py [hunk 24]
```diff
@@ -0,0 +1,323 @@
...
+        #printing message about guesses remaining
+        print_guesses_remaining(letters_guessed, guesses_remaining)
+
+        #asking for new letter
+        letter = input("Please guess a letter: ").lower()
+        
+        #checking if this is really letter, not another symbol, then update remaining warnings and guesses
+        temp_tuple = letter_checker(letter, letters_guessed, guesses_remaining, warnings_remaining, hints_on)
+
+        guesses_remaining, warnings_remaining, letter_is_correct, msg = temp_tuple
```
Comment (Hozhyi) [inline hangman/hangman.py:298]: I'd tried to shorten it a bit differ:
```
guesses_remaining, warnings_remaining, letter_is_correct, msg =
    letter_checker(letter, letters_guessed, guesses_remaining, warnings_remaining, hints_on)
```

File: hangman/hangman.py [hunk 25]
```diff
@@ -0,0 +1,323 @@
...
+    while guesses_remaining > 0 and not is_word_guessed(secret_word, letters_guessed):
+
+        #printing message about guesses remaining
+        print_guesses_remaining(letters_guessed, guesses_remaining)
+
+        #asking for new letter
+        letter = input("Please guess a letter: ").lower()
+        
+        #checking if this is really letter, not another symbol, then update remaining warnings and guesses
+        temp_tuple = letter_checker(letter, letters_guessed, guesses_remaining, warnings_remaining, hints_on)
```
Comment (Hozhyi) [inline hangman/hangman.py:296]: You could check whether user asks hints first and `continue` if he does, then you won't need check it inside `letter_checker` function and won't need additional parameter `hints_on`

File: hangman/hangman.py [hunk 26]
```diff
@@ -0,0 +1,323 @@
...
+        elif not letter_is_correct:
+            print(msg.format(warnings_remaining, get_guessed_word(secret_word, letters_guessed)))
+
+        #if all right - trying to guess
+        else:
+            guesses_remaining, letters_guessed, msg = guessing(secret_word, letter, letters_guessed, guesses_remaining)
+            print(msg.format(get_guessed_word(secret_word, letters_guessed)))
+        
+
+
```
Comment (Hozhyi) [inline hangman/hangman.py:315]: Have you read PEP recommendations?

**Review by Hozhyi — CHANGES_REQUESTED** [review]

**Comment by dmytrochornyy** [comment]
Hi, @Hozhyi ! Did all you said, could you check one more time please?)

File: hangman/hangman.py [hunk 27]
```diff
@@ -0,0 +1,330 @@
...
+    if len(my_word) != len(other_word):
+        return False
+    
+    #if len is same
+    for i in range(len(my_word)):
+        #check if letters on same positions are different (except positions with"_"
+        # and check if we have similar case as: comparing a_bc and abbc)
+        if (my_word[i] != "_") and (my_word[i] != other_word[i]):
+            return False
+        elif (my_word[i] == "_") and (other_word[i] in my_word):
```
Comment (Hozhyi) [inline hangman/hangman.py:246]: You use that "_" symbol so often, that it also makes sense to create a constant for it

File: hangman/hangman.py [hunk 28]
```diff
@@ -0,0 +1,330 @@
...
+
+    #adding all of our matching words in cycle
+    for word_from_list in wordlist:
+
+        #if word is matching
+        if match_with_gaps(my_word, word_from_list):
+            list_with_mathes.append(word_from_list)
+
+    #print matching words
+    print(*list_with_mathes)
```
Comment (Hozhyi) [inline hangman/hangman.py:275]: According to task description - here should be different output

File: hangman/hangman.py [hunk 29]
```diff
@@ -0,0 +1,330 @@
...
+    secret_word: string, the secret word to guess.
+    hints_on: boolean. True, if you want to play with hints. Default = False
+    guesses_remaining: integer, the number of guesses. Default = 6
+    warnings_remaining: integer, the number of warnings. Default = 3
+    Starts up an interactive game of Hangman.
+    '''
+    #printing hello message
+    print_game_start(secret_word)
+    
+    #create blank letters 0/1-table. 0(False) - if letter is not guessed - 1(True) otherwise
```
Comment (Hozhyi) [inline hangman/hangman.py:291]: Seems that this comment is outdated

**Review by Hozhyi — CHANGES_REQUESTED** [review]

File: hangman/hangman.py [hunk 30]
```diff
@@ -0,0 +1,353 @@
...
+
+    Args:
+        secret_word (string): word to guess
+        letters_guessed (set): set with guessed letters
+        guesses_remaining (integer): How many guesses are remaining
+    """
+    # If word was guessed:
+    if is_word_guessed(secret_word, letters_guessed):
+        print("------------")
+        print("\nCongratulations, you won!",
```
Comment (Hozhyi) [inline hangman/hangman.py:234]: Could you add a separate empty print, apart from adding `\n` special symbols, it's will be more readable

File: hangman/hangman.py [hunk 31]
```diff
@@ -0,0 +1,353 @@
...
+
+        # Printing message about guesses remaining
+        print_guesses_remaining(letters_guessed, guesses_remaining)
+
+        # Asking for new letter
+        letter = input("Please guess a letter: ").lower()
+
+        # Activating hints
+        if letter == "*" and hints_on:
+            print("Possible word matches are: ", end="")
```
Comment (Hozhyi) [inline hangman/hangman.py:328]: Could you include that print into `show_possible_matches`?

File: hangman/hangman.py [hunk 32]
```diff
@@ -0,0 +1,353 @@
...
+    while guesses_remaining > 0 and not is_word_guessed(secret_word, letters_guessed):
+
+        # Printing message about guesses remaining
+        print_guesses_remaining(letters_guessed, guesses_remaining)
+
+        # Asking for new letter
+        letter = input("Please guess a letter: ").lower()
+
+        # Activating hints
+        if letter == "*" and hints_on:
```
Comment (Hozhyi) [inline hangman/hangman.py:327]: I believe it makes sense to make this "*" symbol as constant as well

**Review by Hozhyi — CHANGES_REQUESTED** [review]

**Review by Hozhyi — APPROVED** [review]

**Pull request merged**